### PR TITLE
Prepare package to publish on the NPM

### LINKED ISSUE #7: Publish the package in the npm
Similar to the React [issue](https://github.com/ckeditor/ckeditor5-react/issues/5), we need to:

* [x] make sure the README makes sense and corresponds to the code,
* [ ] clearly state in the README that this is a dev preview,
* [ ] create a basic guide in [docs.ckeditor.com](https://docs.ckeditor.com/ckeditor5/latest/builds/index.html); a few words + redirect to the GitHub repo would do the trick. I think maintaining a proper guide before the stabilization is a chore, let's wait for the feedback from the community first,
* [ ] update [FAQ in docs.ckeditor.com](https://docs.ckeditor.com/ckeditor5/latest/builds/guides/faq.html#how-to-use-ckeditor-5-with-frameworks-angular-react-etc),
* [x] a few [tests](https://github.com/ckeditor/ckeditor5-angular/issues/8) would be great,
* [x] some [basic CI](https://github.com/ckeditor/ckeditor5-angular/issues/3) would also help
* [x] add npm script that generates the changelog
* [x] make sure that `npm run publish` will do what we want. And check if we need to use [ckeditor5-dev-env/lib/release-tools/utils/releaserepository.js](https://github.com/ckeditor/ckeditor5-dev/blob/0acc91d27829a88de541519d950284ec870e5321/packages/ckeditor5-dev-env/lib/release-tools/utils/releaserepository.js) after building, like [here](https://github.com/ckeditor/ckeditor5-react/pull/14/files/17ef4105ba36323aa860b8736ea8b40d085b49e2#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R27)
* [x] make sure that the `package.json` created for the `NPM` release will contain only peer angular@6 dependencies

cc @ma2ciek @Reinmar

### LINKED ISSUE #12: npm run e2e fails
Because of missing `e2e/protractor.conf.js` file.

image

### PR DESCRIPTION
**Pull request by ma2ciek**
Other: Prepared package to be published on the NPM. Closes #7. Closes #12.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -37,8 +37,8 @@ testem.log
 /typings
```
Comment (pomek) [inline .gitignore]: In this file - L13 - L27 should be removed. IDE and editors files should be ignored by user's gitignore, not the package.

File: .travis.yml [hunk 2]
```diff
@@ -17,7 +17,9 @@ before_install:
 install:
 - npm install
 script:
-- npm t
+- npm run coverage
+- npm run lint
+- npm run e2e
```
Comment (pomek) [inline .travis.yml]: I would connect these things together. `npm run ... && npm run ...` 

Also, `lint` should be first. If linter fails, the next ones should not be executed.

File: README.md [hunk 3]
```diff
@@ -207,6 +207,12 @@ To run e2e tests run:
 npm run e2e
 ```
 
+To run coverage tests run:
+
+```
```
Comment (pomek) [inline README.md]: \`\`\`bash

File: e2e/protractor.conf.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+const { SpecReporter } = require( 'jasmine-spec-reporter' );
+
+exports.config = {
+	allScriptsTimeout: 30000,
+	specs: [
+		'./src/**/*.e2e-spec.ts'
+	],
+	capabilities: {
+		'browserName': 'chrome'
```
Comment (pomek) [inline e2e/protractor.conf.js]:
```diff
- 'browserName':
+ browserName:
```

File: lib/ckeditor.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
+/**
+ * Base typings for the CKEditor5 elements.
+ */
+export namespace CKEditor5 {
+	/**
+	 * CKEditor5 editor constructor.
+	 */
+	export interface EditorConstructor {
```
Comment (pomek) [inline lib/ckeditor.ts]: `EditorConstructor` => `EditorCreator` or `EditorFactory`?

File: lib/ckeditor.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+		off(): void;
+		stop(): void;
+	}
+
+	/**
+	 * Base Editor class.
+	 *
+	 * See https://docs.ckeditor.com/ckeditor5/latest/api/module_core_editor_editor-Editor.html
+	 */
+	export interface BaseEditor {
```
Comment (pomek) [inline lib/ckeditor.ts]: `EditorInstance` or even `Editor` will be good enough.

File: lib/ckeditor.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+		stopListening( emitter: any, eventName: string, callback: Function ): void;
+		t( ...args: any[] ): void;
+
+		[ property: string ]: any;
+	}
+
+	/**
+	 * Editor that implements `setData()` and `getData()`, e.g. the `ClassicEditor`, `InlineEditor`, etc.
+	 */
+	export interface Editor extends BaseEditor {
```
Comment (pomek) [inline lib/ckeditor.ts]: These methods are provided by the `DataApi` interface (https://docs.ckeditor.com/ckeditor5/latest/api/module_core_editor_utils_dataapimixin-DataApi.html)

I would call it as in the docs then `BaseEditor` could implement the `DataApi` interface.

**Comment by pomek** [comment]
One more thing  - `@Output #change` event calls the function with an object that contains three things:

```ts
this.change.emit( { evt, editor, data } );
```

Do we really need the `data` here? If the editor is available, we can get the data just by calling `editor.getData()`. The same way we did in React integration. 

PS: `evt`? It can be used as a function argument but the object property should be `event`.

```js
editor.on( 'example', evt => {
  evt.stop();
} );

// but:

const evt = new Event( /* ... */ );
editor.fires( 'example', { event: evt, editor: mySuperEditor } );
```

File: lib/ckeditor.ts (hunk 7 above)
Comment (ma2ciek) [reply lib/ckeditor.ts]: So how you would call editors that implement these methods?

File: .gitignore (hunk 1 above)
Comment (ma2ciek) [reply .gitignore]: IMO if somebody is used to the angular environment, then he is used to these `.gitignore` files generated by the `angular cli`.

Comment (pomek) [reply .gitignore]: https://gist.github.com/subfuzion/db7f57fff2fb6998a16c

**Comment by ma2ciek** [comment]
> There is no dist/ in the repository. If someone would install the package from Github, he must build it manually before starts using it.

Dist is huge. there's no way to make publish it on the GH too.

> lib and src are confusing. In fact, important is the lib/ directory which contains the core of the package. I would rename src to demo or sample (whatever that shows the difference between these two directories)

I agree. I'll think on better names

> CKEditor component should accept the editor property instead of build. You want to initialize an editor, not a build.

The `editor` property means for me more an instance than the class.

> There are two package.jsons which will make noise. One of them will be published, the second one... just will be.

It's a common approach for libs using `ng-packagr`.

> .github/PULL_REQUEST_TEMPLATE.md would be welcome if we're going to use the changelog generator,

Fixed.

> One more thing - @Output #change event calls the function with an object that contains three things:

> this.change.emit( { evt, editor, data } );
Do we really need the data here? If the editor is available, we can get the data just by calling editor.getData(). The same way we did in React integration.

It makes it just easier to use. 90% will just use the data property and won't care about the rest params.

> PS: evt? It can be used as a function argument but the object property should be event.

Right, I'll rename it.

**Comment by wwalc** [comment]
> Why GPL+2? cc: @wwalc

Let's stick with GPL 2+ for now. It's always easier to switch to MIT later on than from MIT to GPL.
MIT could be misleading/confusing for outsiders.

File: lib/ckeditor.ts (hunk 7 above)
Comment (ma2ciek) [reply lib/ckeditor.ts]: I'd stick to the CKEditor5 namings as much as possible. That's why I don't like this proposal.

File: lib/ckeditor.ts (hunk 5 above)
Comment (ma2ciek) [reply lib/ckeditor.ts]: Typescript provides many such interfaces for the ES classes, e.g. the `StringConstructor`, etc. It's a common practice IMO.

Comment (pomek) [reply lib/ckeditor.ts]: If it's a common practice then ok… 

So the description of the interface could explain what it does. E.g.: "_A factory that creates a new instance of the editor._"

At this moment I can see the same description in docs and name of the interface which means that the docs could be skipped.

File: lib/ckeditor.ts (hunk 7 above)
Comment (pomek) [reply lib/ckeditor.ts]: > This is a class that implements `DataApi` interface (behavior) and contains the editor's properties (hence `EditorProperties` interface).

**Comment by pomek** [comment]
> > CKEditor component should accept the editor property instead of build. You want to initialize an  editor, not a build.
> 
> The `editor` property means for me more an instance than the class.

As I said, you want to initialize an editor, not a build. The build is a specific kind of the editor but it's still the editor.

`editor` property should be an editor class which will be created. The editor instance could be saved as `editorInstance` or `_editor`. This one is much harder then I had in React because I had available properties as `this.props`. Angular mixes properties required for component and properties which were created when the component was initialized (Am I right?). 

That's a good reason why properties for components should be public and properties created by the component should be private.

> > this.change.emit( { evt, editor, data } );
> >Do we really need the data here? If the editor is available, we can get the data just by calling editor.getData(). The same way we did in React integration.
> 
> It makes it just easier to use. 90% will just use the data property and won't care about the rest params.

If the editor's data could be received using `editor` variable, I would like to remove `data` property. A developer can call `editor.getData()`.

**Comment by pomek** [comment]
Coverage does not work:

image

The whole build should fail if Karma cannot load something.

**Comment by pomek** [comment]
https://david-dm.org/ckeditor/ckeditor5-angular?type=dev

Could we upgrade out-of-date packages?

**Comment by ma2ciek** [comment]
> The whole build should fail if Karma cannot load something.

That's rather a Karma issue.

> When one of the build commands returns a non-zero exit code, the Travis CI build runs the subsequent commands as well, and accumulates the build result.

**Comment by pomek** [comment]
Seems to be an issue of npm:

image

**Comment by ma2ciek** [comment]
> editor property should be an editor class which will be created. The editor instance could be saved as editorInstance or _editor. This one is much harder then I had in React because I had available properties as this.props. Angular mixes properties required for component and properties which were created when the component was initialized (Am I right?).

Components communicate via `@Input` and `@Output` decorators with a template, so providing an interface for communication is not a problem. They share the same scope other properties but they won't be accessible outside (despite tests, so that's why the editor instance isn't private in that component).

> That's a good reason why properties for components should be public and properties created by the component should be private.

And they are, look at the code :)

**Comment by ma2ciek** [comment]
> Could we upgrade out-of-date packages?

Upgraded.

**Comment by ma2ciek** [comment]
Now, there's a coverage threshold anyway, so code without 100% CC won't pass.

**Pull request merged**